File: python/basics/sliding_window/1004_max_consecutive_ones.py
'''
https://leetcode.com/problems/max-consecutive-ones-iii/
sliding windows

'''

def max_consecutive_ones(nums: list[int], k: int):
    l = 0
    N=len(nums)
    max_ones = 0
    zeros = 0
    for r in range(N):
        if nums[r] == 0:
            zeros += 1
        while zeros > k and l < N:
            if l < N and nums[l] == 0:
                zeros -= 1
            l += 1
        window = r-l+1
        max_ones = max(max_ones, window)
    return max_ones




nums = [1,1,1,0,0,0,1,1,1,1,0]
k = 2
#Output: 6
result =  max_consecutive_ones(nums, k)
print("result: " + str(result))

nums = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1]
k = 3
#Output: 10
result =  max_consecutive_ones(nums, k)
print("result: " + str(result))

File: problem_solving/sliding_window/cpp/1004_max_consecutive_ones.cpp
/*'''
https://leetcode.com/problems/max-consecutive-ones-iii/
sliding windows

'''
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"


int max_consecutive_ones(std::vector<int>& nums, int k) {
    int len = nums.size();
    int zeros = 0;
    int max_len = 0;
    int l = 0;
    for(int r = 0; r < len; ++r) {
        if(nums[r] == 0) {
            zeros++;
        }
        while(zeros > k) {
            if(nums[l] == 0) {
                zeros--;
            }
            l++;
        }
        max_len = std::max(max_len, r-l+1);
    }
    return max_len;
}




int main() {
    std::vector<int> nums = {1,1,1,0,0,0,1,1,1,1,0};
    int k = 2;
    //#Output: 6
    int result =  max_consecutive_ones(nums, k);
    std::cout << "result: " << result << std::endl;
    
    nums = {0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1};
    k = 3;
    //#Output: 10
    result =  max_consecutive_ones(nums, k);
    std::cout << "result: " << result << std::endl;
}
